### DISCUSSION TITLE
Accepting selection and executing command with a single return

### DISCUSSION THREAD

**Discussion by cangermueller** [hrsh7th/nvim-cmp#1803] [Q&A] [answered]
Hi,

When selecting an item from the completion menu when pressing return (<cr>) the 1st time, the item is accepted and the menu closes, and a 2nd <cr> is needed to execute the command.

Before <cr>:
<img width="158" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/4807859/af31d95a-959f-4122-9ae5-f98337b62046">

After hitting <cr> the 1st time (the selection is accepted but the file is not yet opened):
<img width="225" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/4807859/9532bc7e-151f-4995-a01e-64da9c4e2b69">

After hitting <cr> the 2nd time, the the file is opened.

Can commands with a single argument (e.g. :e or :tabedit) be executed right away (with a single <cr>)?

This is my config:
<img width="267" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/4807859/50c5a65c-d0e5-4f96-8761-bb726af3e14b">

Thanks!

**Comment by fitrh** [comment]
You can add something like this

```lua
mapping = cmp.mapping.preset.cmdline({
  ['<C-y>'] = {
    c = function()
      if cmp.visible() then
        cmp.confirm({ select = true })
      end

      local cr = vim.api.nvim_replace_termcodes('<CR>', true, true, true)
      vim.api.nvim_feedkeys(cr, 'n', true)
    end,
  },
})
```

Which makes Ctrl+y select the completion item if any and then sends Enter which executes the command

**Comment by cangermueller** [comment, marked as answer]
Great, this worked! Just two small typos (replact -> replace; ce -> cr):
```
mapping = cmp.mapping.preset.cmdline({
  ['<CR>'] = {
    c = function()
      if cmp.visible() then
        cmp.confirm({ select = true })
      end

      local cr = vim.api.nvim_replace_termcodes('<CR>', true, true, true)
      vim.api.nvim_feedkeys(cr, 'n', true)
    end,
  },
})
```

In the meantime, I found that the desired behavior can also be obtained with `<c-j>`. `<c-y>` accepts the selection and does not execute by default.

**Reply by cangermueller** [reply]
The drawback with the <CR> mapping above is that the first item of the completion menu is also executed if it is not selected.

For example, hitting <CR> after entering `:w` does not execute `:w`, but `:wq` if the the first unselected item in the menu is `wq` 
<img width="113" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/4807859/201cb9aa-af35-4d2b-bfae-0a4a8d353b68">


How can I check if an item of the menu is actually selected and only execute it in this case? And what is entered in the command line otherwise? Basically wrapping the command above in  some if statement.

**Reply by fitrh** [reply]
Set `select = false`


```lua
cmp.confirm({ select = false })
```

**Reply by cangermueller** [reply]
Thanks, this works!

**Comment by esn89** [comment]
@cangermueller how did you get your pop up menu to display only the available commands?

Mine shows "variable" on the rightside of it:

[see here](https://imgur.com/a/k27kVhY)

**Comment by cangermueller** [comment]
Maybe my formatting options are responsible for this? See my config below.

```
  {
    'hrsh7th/cmp-cmdline',
    dependencies = {
      'hrsh7th/nvim-cmp',
    },
    enabled = false, -- Required because of nvgoog
    keys = { '/', '?', ':' },
    opts = {
      formatting = {
        format = function(_, vim_item)
          -- Avoid truncating labels
          vim_item.menu = ''
          vim_item.kind = ''
          return vim_item
        end,
      },
    },
    config = function(_, opts)
      local cmp = require('cmp')
      cmp.setup(opts)

      -- Select and execute on first <CR>. <C-y> selects only.
      local cmdline_mapping = cmp.mapping.preset.cmdline({
        ['<CR>'] = {
          c = function()
            if cmp.visible() then
              cmp.confirm({ select = false })
            end

            local cr =
              vim.api.nvim_replace_termcodes('<CR>', true, true, true)
            vim.api.nvim_feedkeys(cr, 'n', true)
          end,
        },
      })

      cmp.setup.cmdline({ '/', '?' }, {
        mapping = cmdline_mapping,
        sources = {
          { name = 'buffer' },
        },
      })

      cmp.setup.cmdline(':', {
        mapping = cmdline_mapping,
        sources = cmp.config.sources({
          { name = 'path' },
          {
            name = 'cmdline',
          },
        }),
      })
    end,
  },
```

**Reply by esn89** [reply]
That was it! 

Thank you @cangermueller
